src/Transform/ONNX/ConstProp.cpp

```diff
@@ -781,79 +720,35 @@ Value ConstPropCast(
 
 //===----------------------------------------------------------------------===//
 // Code to perform constant propagation for SliceOp.
-//
-// TODO: Move this to a slice method in ElementsAttrBuilder.
 //===----------------------------------------------------------------------===//
 
-void ConstPropSliceImpl(ShapedType outputType,
-    const ONNXSliceOpShapeHelper &shapeHelper, ElementsAttr inputElements,
-    MutableArrayRef<WideNum> outputData) {
-  size_t rank = outputType.getRank();
-  auto outputShape = outputType.getShape();
-  std::vector<int64_t> inputStrides =
-      getStrides(inputElements.getType().getShape());
-  ArrayBuffer<WideNum> inputBuffer = getElementsWideNums(inputElements);
-  const WideNum *start = inputBuffer.get().begin();
-  SmallVector<int64_t, 4> steps(rank, 0);
-  for (size_t axis = 0; axis < rank; ++axis) {
-    start += shapeHelper.starts[axis].getLiteral() * inputStrides[axis];
-    steps[axis] = shapeHelper.steps[axis].getLiteral() * inputStrides[axis];
-  }
-  for (auto &idxpos : StridesRange<1>(outputShape, {steps}))
-    outputData[idxpos.flattenedIndex] = *(start + idxpos[0]);
-}
-
 Value ConstPropSlice(
     PatternRewriter &rewriter, Value replacingValue, Value constValue) {
   ConstPropCounters::count("Slice", {constValue});
   Operation *op = replacingValue.getDefiningOp();
-  ONNXSliceOp sliceOp = cast<ONNXSliceOp>(op);
 
-  // Get starts, ends, axes and steps via ShapeHelper.
+  // Get shape, starts, steps via ShapeHelper.
   ONNXSliceOpShapeHelper shapeHelper(op, {});
-  if (failed(shapeHelper.computeShape())) {
-    sliceOp.emitError("Failed to scan " + ONNXSliceOp::getOperationName() +
-                      " parameters successfully");
-    return nullptr;
-  }
+  auto outcome = shapeHelper.computeShape();
+  assert(succeeded(outcome) && "Failed to scan slice op parameters");
+  auto toLiterals = [](ArrayRef<IndexExpr> ies) {
+    return llvm::map_range(ies, [](IndexExpr ie) { return ie.getLiteral(); });
+  };
+  SmallVector<int64_t> shape(toLiterals(shapeHelper.getOutputDims()));
+  SmallVector<int64_t> starts(toLiterals(shapeHelper.starts));
+  SmallVector<int64_t> steps(toLiterals(shapeHelper.steps));
```

Comment:
You can use `IndexExpr`, e.g.
```
SmallVector<int64_t> shape;
IndexExpr::getLiteral(shapeHelper.getOutputDims(), shape);
```

Reply:
done, thanks for the suggestion

I noticed that the getLiteral() implementation was missing a `const` qualifier for the first argument and went ahead and made IndexExpr::getLiteral/Shape/DynSymbols/OpOrFoldResults use ArrayRef like getValues

Reply:
Thanks!